cluster/context.go

```diff
@@ -25,13 +25,102 @@ import (
 // goes wrong during the business logic execution, database objects can be
 // rolled back.
 type Context struct {
-	*sql.Tx
-	Main context.Context
+	// tenant in question
+	TenantShortName string
+	tenantTx        *sql.Tx
+	tenantDB        *sql.DB
+	mainTx          *sql.Tx
+	ControlCtx      context.Context
+	// a user identifier of who is starting the context
+	WhoAmI string
 }
 
-// Creates a new `Context` given an initial transaction and `context.Context`
+// MainTx returns a transaction against the Main DB, if none has been started, it starts one
+func (c *Context) MainTx() (*sql.Tx, error) {
+	if c.mainTx == nil {
+		db := GetInstance().Db
+		tx, err := db.BeginTx(c.ControlCtx, nil)
+		if err != nil {
+			return nil, err
+		}
+		c.mainTx = tx
+	}
+	return c.mainTx, nil
+}
+
+// TenantDB returns a configured DB connection for the Tenant DB
+func (c *Context) TenantDB() *sql.DB {
+	if c.tenantDB == nil {
+		db := GetInstance().GetTenantDB(c.TenantShortName)
+		c.tenantDB = db
+	}
+	return c.tenantDB
+}
+
+// TenantTx returns a transaction against the Tenant DB, if none has been started, it starts one
+func (c *Context) TenantTx() (*sql.Tx, error) {
+	if c.mainTx == nil {
+		db := c.TenantDB()
+		tx, err := db.BeginTx(c.ControlCtx, nil)
+		if err != nil {
+			return nil, err
+		}
+		c.tenantTx = tx
+	}
+	return c.tenantTx, nil
+}
+
+// Commit commits the any transaction that was started on this context
+func (c *Context) Commit() error {
+	// commit tenant schema tx
+	if c.tenantTx != nil {
+		err := c.tenantTx.Commit()
+		if err != nil {
+			return err
+		}
+		// restart the txn
+		c.tenantTx = nil
+	}
+	// commit main schema tx
+	if c.mainTx != nil {
+		err := c.mainTx.Commit()
+		if err != nil {
+			return err
+		}
+		// restart the txn
+		c.mainTx = nil
+	}
+	return nil
+}
+
+func (c *Context) Rollback() error {
```

Comment:
if we want to just rollback the TenanTx this will rollback both mainTx and Tenantx or are they mutually exclusive?

Reply:
you can get each transaction individually and roll it back